src/state/learnocaml_data.ml

```diff
@@ -513,6 +513,57 @@ module Exercise = struct
       | Some a -> a
       | None -> a.default
 
+  (* Open/Close invariants:
+     forall exo_status,
+     1.(REQUIRED):
+        (exo_status.assignments.default <> Open /\ Token.Map.for_all (fun _ st -> st <> Open))
```

Comment:
I have some troubles understanding this formula because of the number of cases. For instance, it looks like you can have `exo_status.assignments.default = Close` and some token status equals to `Assigned x` for some `x`. Is that correct?

What is the type of `st` here?

Do you really want that, or:
```
forall exo_status e, exists x, e.assignments.default <> Assigned x -> forall known token t, t = e.assignments.default?
```

Reply:
As we said in the telco, this is mostly a (map + default value) to emulate a constant function except for specific tokens.

We just don't want to have at the same time Open and Closed states, so that there is only 4 possible global states.